File: Assets/Scripts/GnatEnemy.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class GnatEnemy : MonoBehaviour
{

    // RB and SR components.
    Rigidbody rb = null;
    SpriteRenderer sr;

    // "Gravity" force towards the player.
    [SerializeField] float gravity_scale = 1.0f;

    // All different gnat sprites.
    [SerializeField] Sprite[] sprites;
    
    // Current velocity.
    Vector2 cur_vel = Vector2.zero;

    // When enabled:
    private void OnEnable()
    {
        // If not initialized, do so. Otherwise, set the velocity.
        if (rb == null) Init();
        else rb.velocity = cur_vel;

        // Run the sprite coroutine.
        StartCoroutine(ChangeSprite());

    }

    void Init()
    {
        // Get the components.
        sr = GetComponent<SpriteRenderer>();
        rb = GetComponent<Rigidbody>();

        // Pick a random angle to fly from the egg, and set the velocity accordingly.
        float angle = Random.Range(0.0f, Mathf.PI*2);
        rb.velocity = new Vector2(Mathf.Cos(angle) * 2.0f, Mathf.Sin(angle) * 2.0f);

        // Modify the drag to be 0.1.
        rb.drag = 0.1f;
    }

    void Update()
    {
        // Store the current velocity.
        cur_vel = rb.velocity;

        // Wrap the movement around the bottom and left side of the screen.
        // Maintains velocity.
        if (transform.position.y < 0.0f)
        {
            transform.position = new Vector3(0.0f, transform.position.x, 0.0f);
            rb.velocity = new Vector3(-rb.velocity.y, rb.velocity.x);
        }
        else if (transform.position.x < 0.0f)
        {
            transform.position = new Vector3(transform.position.y, 0.0f, 0.0f);
            rb.velocity = new Vector3(rb.velocity.y, -rb.velocity.x);
        }

        // Calculate new gravity force based on the angle to the player and apply it.
        float angle = Mathf.Abs(Mathf.Atan(transform.position.y / transform.position.x));
        Vector2 gravity = new Vector2(Mathf.Sign(transform.position.x) * Mathf.Cos(angle) * -gravity_scale, Mathf.Sign(transform.position.y) * Mathf.Sin(angle) * -gravity_scale);
        rb.AddForce(gravity);
    }

    // Sprite changing coroutine. Randomly change sprites every half second.
    IEnumerator ChangeSprite()
    {
        while (true)
        {
            yield return new WaitForSeconds(0.5f);
            int sprite_num = (int)Random.Range(0, 3.999f);
            sr.sprite = sprites[sprite_num];
        }
    }

}

File: component_types/GnatEnemy.lua
GnatEnemy = {

    velocity = Vector2(0,0),
    cur_frame = 0,
    gravity_scale = 0.0166666,

    OnEnable = function (self)
        self.rb2d:SetVelocity(self.velocity)
    end,

    OnDisable = function (self)
        self.rb2d:SetVelocity(Vector2(0,0))
    end,

    OnStart = function (self)
        self.sr = self.actor:GetComponent("SpriteRenderer")
        self.rb2d = self.actor:GetComponent("Rigidbody2D")
        local angle = math.random() * 2 * math.pi
        self.velocity = Vector2(math.cos(angle) * 2, math.sin(angle) * 2)
        self.rb2d:SetVelocity(self.velocity)
    end,

    OnUpdate = function (self)
        self.velocity = self.rb2d:GetVelocity()
        local position = self.rb2d:GetPosition()
        if position.y < 0 then
            position.y = position.x
            position.x = 0
            self.rb2d:SetPosition(position)
            self.rb2d:SetVelocity(Vector2(-self.velocity.y,self.velocity.x))
        elseif position.x < 0 then
            position.x = position.y
            position.y = 0
            self.rb2d:SetPosition(position)
            self.rb2d:SetVelocity(Vector2(self.velocity.y,-self.velocity.x))
        end
        local angle = math.abs(math.atan(position.y,position.x))
        local x_sign = 1
        if position.x < 0 then
            x_sign = -1
        end
        local y_sign = 1
        if position.y < 0 then
            y_sign = -1
        end
        local gravity = Vector2(x_sign*math.cos(angle)*-self.gravity_scale,y_sign*math.cos(angle)*-self.gravity_scale)
        self.rb2d:AddForce(gravity)
    end,

    OnLateUpdate = function (self)
        self.cur_frame = self.cur_frame + 1
        if self.cur_frame % 30 == 0 then
            self.sr.sprite = "enemies/gnat/gnat_" .. tostring(math.random(0,3))
        end
    end

}